lib/improver/spotdata_new/neighbour_finding.py

```diff
@@ -0,0 +1,488 @@
+        self.geodetic_coordinate_system = (
+            orography.coord_system().as_cartopy_crs().is_geodetic())
+
+        # Enforce x-y coordinate order for input cubes.
+        orography = enforce_coordinate_ordering(
+            orography, [orography.coord(axis='x').name(),
+                        orography.coord(axis='y').name()])
+        land_mask = enforce_coordinate_ordering(
+            land_mask, [land_mask.coord(axis='x').name(),
+                        land_mask.coord(axis='y').name()])
+
+        # Remap site coordinates on to coordinate system of the model grid.
+        site_x_coords = np.array([site[self.site_x_coordinate]
+                                  for site in sites])
+        site_y_coords = np.array([site[self.site_y_coordinate]
+                                  for site in sites])
+        site_coords = self._transform_sites_coordinate_system(
+            site_x_coords, site_y_coords, orography)
+
+        # Exclude any sites falling outside the domain given by the cube and
+        # notify the user.
+        sites, site_coords, site_x_coords, site_y_coords = (
+            self.check_sites_are_within_domain(
+                sites, site_coords, site_x_coords, site_y_coords, orography))
+
+        # Find nearest neighbour point using quick iris method.
+        nearest_indices = self.get_nearest_indices(site_coords, orography)
+
+        # Create an array containing site altitudes, using the nearest point
+        # orography height for any that are unset.
+        site_altitudes = np.array([site.get(self.site_altitude, None)
+                                   for site in sites])
+        site_altitudes = np.where(np.isnan(site_altitudes.astype(float)),
+                                  orography.data[tuple(nearest_indices.T)],
+                                  site_altitudes)
+
+        # If further constraints are being applied, build a KD Tree which
+        # includes points filtered by constraint.
+        if self.land_constraint or self.minimum_dz:
+            tree, index_nodes = self.build_KDTree(land_mask)
+
+            # Site coordinates made cartesian for global coordinate system
+            if self.geodetic_coordinate_system:
+                site_coords = self.geocentric_cartesian(
+                    orography, site_coords[:, 0], site_coords[:, 1])
+
+            if not self.minimum_dz:
+                distances, node_indices = tree.query([site_coords])
+                land_neighbour_indices, = index_nodes[node_indices]
+                distances = np.array([distances[0], distances[0]]).T
+                nearest_indices = np.where(distances < self.search_radius,
+                                           land_neighbour_indices,
+                                           nearest_indices)
+            else:
+                distances, node_indices = tree.query(
+                    [site_coords], distance_upper_bound=self.search_radius,
+                    k=36)
```

Comment:
No. I envisaged this as a sensible maximum, something like a 6x6 neighbourhood (though it may not be a square 6x6 depending on the grid). It's very unlikely that going much further than this would be appropriate for spotdata unless it was a very small spatial scale grid (e.g. at the poles).